{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from functools import reduce\n",
    "from cyvcf2 import VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read Coding coordinates\n",
    "gene_regions = pd.read_csv(\"/gymreklab-tscc/helia/ensembl/experiments/coding_regions/coding_regions_sorted_corrected.bed\", \n",
    "                           sep=\"\\t\", header=None)\n",
    "gene_regions.columns = ['chrom', 'start', 'end', 'name', 'x', 'strand']\n",
    "\n",
    "# Read TR coordinates\n",
    "TR_regions = pd.read_csv(\"/gymreklab-tscc/helia/ensembl/experiments/coding_regions/all_TR_coordinates.bed\",\n",
    "                         sep=\"\\t\", header=None)\n",
    "TR_regions.columns = ['chrom', 'start', 'end']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_input(chrom):\n",
    "    gene_chrom = gene_regions[gene_regions['chrom'] == chrom]\n",
    "    gene_chrom = gene_chrom.drop_duplicates(subset = ['chrom', 'start', 'end'])\n",
    "    \n",
    "    TR_chrom = TR_regions[TR_regions['chrom'] == chrom]\n",
    "    TR_chrom = TR_chrom.drop_duplicates(subset = ['chrom', 'start', 'end'])\n",
    "    \n",
    "    with open(f\"inputs/input_{chrom}.txt\", \"w\") as f:\n",
    "        f.write(f\"{len(gene_chrom)} {len(TR_chrom)}\")\n",
    "        f.write(\"\\n\")\n",
    "        for index,row in gene_chrom.iterrows():\n",
    "            f.write(f\"{row['start']} {row['end']}\")\n",
    "            f.write(\"\\n\")\n",
    "            \n",
    "        for index,row in TR_chrom.iterrows():\n",
    "            f.write(f\"{row['start']} {row['end']}\")\n",
    "            f.write(\"\\n\")  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Writing TR and coding regions to input file ####\n",
    "for chrom in range(1,23):\n",
    "    create_input(f\"chr{chrom}\")   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### calling overlap_finder.cpp to find overlap and write them to output ####\n",
    "!g++ overlap_finder.cpp --std c++11\n",
    "for chrom in range(1,23):\n",
    "    input_addr = f\"inputs/input_chr{chrom}.txt\"\n",
    "    output_addr = f\"inputs/overlap_chr{chrom}.txt\"\n",
    "    !./a.out < $input_addr > $output_addr\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/site-packages/ipykernel_launcher.py:7: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "  import sys\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TR_end</th>\n",
       "      <th>TR_start</th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>960798</td>\n",
       "      <td>960770</td>\n",
       "      <td>chr1</td>\n",
       "      <td>960800</td>\n",
       "      <td>960693</td>\n",
       "      <td>id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>970857</td>\n",
       "      <td>970841</td>\n",
       "      <td>chr1</td>\n",
       "      <td>971006</td>\n",
       "      <td>970685</td>\n",
       "      <td>id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>976199</td>\n",
       "      <td>976178</td>\n",
       "      <td>chr1</td>\n",
       "      <td>976269</td>\n",
       "      <td>976171</td>\n",
       "      <td>id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1232644</td>\n",
       "      <td>1232633</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1233268</td>\n",
       "      <td>1232278</td>\n",
       "      <td>id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1232818</td>\n",
       "      <td>1232807</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1233268</td>\n",
       "      <td>1232278</td>\n",
       "      <td>id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>50603361</td>\n",
       "      <td>50603316</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50603498</td>\n",
       "      <td>50603222</td>\n",
       "      <td>id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>50604160</td>\n",
       "      <td>50604149</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50605064</td>\n",
       "      <td>50603840</td>\n",
       "      <td>id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>50604573</td>\n",
       "      <td>50604544</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50605064</td>\n",
       "      <td>50603840</td>\n",
       "      <td>id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>50604557</td>\n",
       "      <td>50604546</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50605064</td>\n",
       "      <td>50603840</td>\n",
       "      <td>id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>50604741</td>\n",
       "      <td>50604704</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50605064</td>\n",
       "      <td>50603840</td>\n",
       "      <td>id</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6492 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       TR_end  TR_start  chrom  gene_end  gene_start  id\n",
       "0      960798    960770   chr1    960800      960693  id\n",
       "1      970857    970841   chr1    971006      970685  id\n",
       "2      976199    976178   chr1    976269      976171  id\n",
       "3     1232644   1232633   chr1   1233268     1232278  id\n",
       "4     1232818   1232807   chr1   1233268     1232278  id\n",
       "..        ...       ...    ...       ...         ...  ..\n",
       "157  50603361  50603316  chr22  50603498    50603222  id\n",
       "158  50604160  50604149  chr22  50605064    50603840  id\n",
       "159  50604573  50604544  chr22  50605064    50603840  id\n",
       "160  50604557  50604546  chr22  50605064    50603840  id\n",
       "161  50604741  50604704  chr22  50605064    50603840  id\n",
       "\n",
       "[6492 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### Load overlaps and merge ####\n",
    "overlaps = pd.DataFrame(columns = ['chrom','TR_start', 'TR_end', 'gene_start', 'gene_end'])\n",
    "for chrom in range(1,23):\n",
    "    overlaps_chrom = pd.read_csv(f\"inputs/overlap_chr{chrom}.txt\", sep = \"\\t\", header = None)\n",
    "    overlaps_chrom.columns = ['TR_start', 'TR_end', 'gene_start', 'gene_end']\n",
    "    overlaps_chrom['chrom'] = \"chr\" + str(chrom)\n",
    "    overlaps = pd.concat([overlaps, overlaps_chrom])\n",
    "overlaps['id'] = \"id\"\n",
    "overlaps = overlaps.drop_duplicates(subset = ['chrom','TR_start', 'TR_end'])\n",
    "overlaps[['TR_start', 'TR_end', 'gene_start', 'gene_end']] = overlaps[['TR_start', 'TR_end', 'gene_start', 'gene_end']].astype(int)\n",
    "overlaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#TRs found by cpp code: 6492, #TRs found by bedtools: 6492, #TRs found by both: 6492\n"
     ]
    }
   ],
   "source": [
    "#### Compare with previous overlaps found with bedtools where TR shouldn't be entirely inside coding region ####\n",
    "\n",
    "old_overlaps = pd.read_csv(\"/gymreklab-tscc/helia/ensembl/experiments/coding_regions/mapping.txt\", sep = \"\\t\", header = None)\n",
    "\n",
    "cnt = 0\n",
    "old_overlaps_entire = []\n",
    "for index, row in old_overlaps.iterrows():\n",
    "    if row[1] >= row[4] and row[2] <= row[5]: # check for entirely being inside the gene\n",
    "        old_overlaps_entire.append(row)\n",
    "old_overlaps_entire = pd.DataFrame(old_overlaps_entire)\n",
    "old_overlaps_entire.columns =  ['chrom','TR_start', 'TR_end','chrom2', 'gene_start', 'gene_end','Gene_name','y','z']\n",
    "old_overlaps_entire = old_overlaps_entire.drop_duplicates(subset = ['chrom','TR_start', 'TR_end'])\n",
    "old_overlaps_entire[['TR_start', 'TR_end', 'gene_start', 'gene_end']] = old_overlaps_entire[['TR_start', 'TR_end', 'gene_start', 'gene_end']].astype(int)\n",
    "compare = pd.merge(old_overlaps_entire, overlaps, left_on = ['chrom', 'TR_start', 'TR_end'], right_on = ['chrom', 'TR_start', 'TR_end'], how = \"outer\")\n",
    "print(f\"#TRs found by cpp code: {len(overlaps)}, #TRs found by bedtools: {len(old_overlaps_entire)}, #TRs found by both: {len(compare)}\")\n",
    "      \n",
    "      \n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_overlaps_entire[['chrom','TR_start','TR_end']].to_csv(\"TR_intersect.txt\", sep = \"\\t\", header=False, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Reading heterozygousity outputs ####\n",
    "\n",
    "AFR_df = pd.DataFrame(columns = [\"chrom\",\"start\",\"end\",\"het-1\"])\n",
    "AMR_df = pd.DataFrame(columns = [\"chrom\",\"start\",\"end\",\"het-1\"])\n",
    "EAS_df = pd.DataFrame(columns = [\"chrom\",\"start\",\"end\",\"het-1\"])\n",
    "SAS_df = pd.DataFrame(columns = [\"chrom\",\"start\",\"end\",\"het-1\"])\n",
    "EUR_df = pd.DataFrame(columns = [\"chrom\",\"start\",\"end\",\"het-1\"])\n",
    "\n",
    "all_df = [EUR_df, SAS_df, EAS_df, AMR_df, AFR_df]\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "pops = ['EUR', 'SAS', 'EAS', 'AMR', 'AFR']\n",
    "\n",
    "for i in range(5):\n",
    "    for chrom in range(1,23):\n",
    "        df = pd.read_csv(f\"/gymreklab-tscc/helia/ensembl/experiments/coding_regions/heterozygosity/coding_het_{pops[i]}_{chrom}.tab\",sep=\"\\t\")\n",
    "        all_df[i] = pd.concat([all_df[i], df])\n",
    "    \n",
    "    all_df[i]['end'] = all_df[i]['end'] - 1\n",
    "    all_df[i] = pd.merge(all_df[i], old_overlaps_entire, left_on = ['chrom', 'start', 'end'], \n",
    "                       right_on = ['chrom', 'TR_start', 'TR_end'], how = \"inner\")\n",
    "    all_df[i] = all_df[i].drop_duplicates(subset = [\"chrom\", \"start\",\"end\"])\n",
    "    all_df[i] = all_df[i][['chrom', 'start', 'end', \"Gene_name\", 'het-1']]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>EUR_het</th>\n",
       "      <th>SAS_het</th>\n",
       "      <th>EAS_het</th>\n",
       "      <th>AMR_het</th>\n",
       "      <th>AFR_het</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>960770</td>\n",
       "      <td>960798</td>\n",
       "      <td>ENST00000338591.8_cds_0_0_chr1_960694_f</td>\n",
       "      <td>0.00321</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>970841</td>\n",
       "      <td>970857</td>\n",
       "      <td>ENST00000379409.6_cds_4_0_chr1_970686_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>976178</td>\n",
       "      <td>976199</td>\n",
       "      <td>ENST00000341290.6_cds_0_0_chr1_976172_r</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1232633</td>\n",
       "      <td>1232644</td>\n",
       "      <td>ENST00000379198.5_cds_0_0_chr1_1232279_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1232807</td>\n",
       "      <td>1232818</td>\n",
       "      <td>ENST00000379198.5_cds_0_0_chr1_1232279_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6487</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50603316</td>\n",
       "      <td>50603361</td>\n",
       "      <td>ENST00000329492.6_cds_2_0_chr22_50603223_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00341</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6488</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604149</td>\n",
       "      <td>50604160</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6489</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604544</td>\n",
       "      <td>50604573</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00684</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6490</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604546</td>\n",
       "      <td>50604557</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6491</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604704</td>\n",
       "      <td>50604741</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00171</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6492 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end                                   gene_name  \\\n",
       "0      chr1    960770    960798     ENST00000338591.8_cds_0_0_chr1_960694_f   \n",
       "1      chr1    970841    970857     ENST00000379409.6_cds_4_0_chr1_970686_f   \n",
       "2      chr1    976178    976199     ENST00000341290.6_cds_0_0_chr1_976172_r   \n",
       "3      chr1   1232633   1232644    ENST00000379198.5_cds_0_0_chr1_1232279_f   \n",
       "4      chr1   1232807   1232818    ENST00000379198.5_cds_0_0_chr1_1232279_f   \n",
       "...     ...       ...       ...                                         ...   \n",
       "6487  chr22  50603316  50603361  ENST00000329492.6_cds_2_0_chr22_50603223_f   \n",
       "6488  chr22  50604149  50604160  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "6489  chr22  50604544  50604573  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "6490  chr22  50604546  50604557  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "6491  chr22  50604704  50604741  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "\n",
       "      EUR_het  SAS_het  EAS_het  AMR_het  AFR_het  \n",
       "0     0.00321  0.00000  0.00000      0.0  0.00233  \n",
       "1     0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "2     0.00000  0.00332  0.00000      0.0  0.00000  \n",
       "3     0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "4     0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "...       ...      ...      ...      ...      ...  \n",
       "6487  0.00000  0.00000  0.00341      0.0  0.00447  \n",
       "6488  0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "6489  0.00000  0.00000  0.00684      0.0  0.00000  \n",
       "6490  0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "6491  0.00000  0.00000  0.00171      0.0  0.00112  \n",
       "\n",
       "[6492 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Merging all pops ####\n",
    "\n",
    "df_all_pop = reduce(lambda  left,right: pd.merge(left,right,on=['chrom','start','end', 'Gene_name'],\n",
    "                                            how='inner'), all_df)\n",
    "df_all_pop.columns = ['chrom', 'start', 'end', 'gene_name', 'EUR_het', 'SAS_het', 'EAS_het', 'AMR_het', 'AFR_het']\n",
    "display(df_all_pop)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Load Motif ####\n",
    "motifs = []\n",
    "for chrom in range(1,23):\n",
    "    vcf = VCF(f\"/gymreklab-tscc/helia/ensembl/experiments/coding_regions/heterozygosity/coding_calls_{chrom}.vcf.gz\")\n",
    "    for record in vcf:\n",
    "        motifs.append([record.CHROM, record.POS, record.INFO['END'], record.INFO['RU']])\n",
    "motifs = pd.DataFrame(motifs, columns = ['chrom', 'start', 'end', 'motif'])\n",
    "motifs = motifs.drop_duplicates(subset = [\"chrom\", \"start\",\"end\"])\n",
    "df_all_pop = pd.merge(df_all_pop, motifs, on = ['chrom', 'start', 'end'])\n",
    "df_all_pop.to_csv(\"Supplementary_Table_7.csv\", index = False, sep = \"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
