{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from functools import reduce\n",
    "from cyvcf2 import VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>TR_start</th>\n",
       "      <th>TR_end</th>\n",
       "      <th>chrom2</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>960770</td>\n",
       "      <td>960798</td>\n",
       "      <td>chr1</td>\n",
       "      <td>960693</td>\n",
       "      <td>960800</td>\n",
       "      <td>ENST00000338591.8_cds_0_0_chr1_960694_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>970841</td>\n",
       "      <td>970857</td>\n",
       "      <td>chr1</td>\n",
       "      <td>970685</td>\n",
       "      <td>971006</td>\n",
       "      <td>ENST00000379409.6_cds_4_0_chr1_970686_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>976178</td>\n",
       "      <td>976199</td>\n",
       "      <td>chr1</td>\n",
       "      <td>976171</td>\n",
       "      <td>976269</td>\n",
       "      <td>ENST00000433179.4_cds_3_0_chr1_976172_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1232633</td>\n",
       "      <td>1232644</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1232278</td>\n",
       "      <td>1233268</td>\n",
       "      <td>ENST00000379198.5_cds_0_0_chr1_1232279_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1232807</td>\n",
       "      <td>1232818</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1232278</td>\n",
       "      <td>1233268</td>\n",
       "      <td>ENST00000379198.5_cds_0_0_chr1_1232279_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20831</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50603316</td>\n",
       "      <td>50603361</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50603222</td>\n",
       "      <td>50603498</td>\n",
       "      <td>ENST00000329492.6_cds_2_0_chr22_50603223_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20832</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604149</td>\n",
       "      <td>50604160</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50603840</td>\n",
       "      <td>50605064</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20833</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604544</td>\n",
       "      <td>50604573</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50603840</td>\n",
       "      <td>50605064</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20834</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604546</td>\n",
       "      <td>50604557</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50603840</td>\n",
       "      <td>50605064</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20835</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604704</td>\n",
       "      <td>50604741</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50603840</td>\n",
       "      <td>50605064</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6074 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom  TR_start    TR_end chrom2  gene_start  gene_end  \\\n",
       "9       chr1    960770    960798   chr1      960693    960800   \n",
       "10      chr1    970841    970857   chr1      970685    971006   \n",
       "11      chr1    976178    976199   chr1      976171    976269   \n",
       "14      chr1   1232633   1232644   chr1     1232278   1233268   \n",
       "15      chr1   1232807   1232818   chr1     1232278   1233268   \n",
       "...      ...       ...       ...    ...         ...       ...   \n",
       "20831  chr22  50603316  50603361  chr22    50603222  50603498   \n",
       "20832  chr22  50604149  50604160  chr22    50603840  50605064   \n",
       "20833  chr22  50604544  50604573  chr22    50603840  50605064   \n",
       "20834  chr22  50604546  50604557  chr22    50603840  50605064   \n",
       "20835  chr22  50604704  50604741  chr22    50603840  50605064   \n",
       "\n",
       "                                        Gene_name  y  z  \n",
       "9         ENST00000338591.8_cds_0_0_chr1_960694_f  0  +  \n",
       "10        ENST00000379409.6_cds_4_0_chr1_970686_f  0  +  \n",
       "11        ENST00000433179.4_cds_3_0_chr1_976172_r  0  -  \n",
       "14       ENST00000379198.5_cds_0_0_chr1_1232279_f  0  +  \n",
       "15       ENST00000379198.5_cds_0_0_chr1_1232279_f  0  +  \n",
       "...                                           ... .. ..  \n",
       "20831  ENST00000329492.6_cds_2_0_chr22_50603223_f  0  +  \n",
       "20832  ENST00000329492.6_cds_4_0_chr22_50603841_f  0  +  \n",
       "20833  ENST00000329492.6_cds_4_0_chr22_50603841_f  0  +  \n",
       "20834  ENST00000329492.6_cds_4_0_chr22_50603841_f  0  +  \n",
       "20835  ENST00000329492.6_cds_4_0_chr22_50603841_f  0  +  \n",
       "\n",
       "[6074 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### Compare with previous overlaps found with bedtools where TR shouldn't be entirely inside coding region ####\n",
    "\n",
    "old_overlaps = pd.read_csv(\"/expanse/projects/gymreklab/helia/ensembl/experiments/coding_regions/mapping.txt\", sep = \"\\t\", header = None)\n",
    "\n",
    "old_overlaps_entire = []\n",
    "for index, row in old_overlaps.iterrows():\n",
    "    if row[1] >= row[4] and row[2] <= row[5]: # check for entirely being inside the gene\n",
    "        old_overlaps_entire.append(row)\n",
    "old_overlaps_entire = pd.DataFrame(old_overlaps_entire)\n",
    "old_overlaps_entire.columns =  ['chrom','TR_start', 'TR_end','chrom2', 'gene_start', 'gene_end','Gene_name','y','z']\n",
    "old_overlaps_entire = old_overlaps_entire.drop_duplicates(subset = ['chrom','TR_start', 'TR_end'])\n",
    "old_overlaps_entire[['TR_start', 'TR_end', 'gene_start', 'gene_end']] = old_overlaps_entire[['TR_start', 'TR_end', 'gene_start', 'gene_end']].astype(int)\n",
    "old_overlaps_entire\n",
    "\n",
    "      \n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_overlaps_entire[['chrom','TR_start','TR_end']].to_csv(\"TR_intersect.txt\", sep = \"\\t\", header=False, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Reading heterozygousity outputs ####\n",
    "\n",
    "AFR_df = pd.DataFrame(columns = [\"chrom\",\"start\",\"end\",\"het-1\"])\n",
    "AMR_df = pd.DataFrame(columns = [\"chrom\",\"start\",\"end\",\"het-1\"])\n",
    "EAS_df = pd.DataFrame(columns = [\"chrom\",\"start\",\"end\",\"het-1\"])\n",
    "SAS_df = pd.DataFrame(columns = [\"chrom\",\"start\",\"end\",\"het-1\"])\n",
    "EUR_df = pd.DataFrame(columns = [\"chrom\",\"start\",\"end\",\"het-1\"])\n",
    "\n",
    "all_df = [EUR_df, SAS_df, EAS_df, AMR_df, AFR_df]\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "pops = ['EUR', 'SAS', 'EAS', 'AMR', 'AFR']\n",
    "for i in range(5):\n",
    "    for chrom in range(1,23):\n",
    "        df = pd.read_csv(f\"/expanse/projects/gymreklab/helia/ensembl/experiments/allele_freq/freqs/freqs_chr{chrom}_{pops[i]}.tab\",\n",
    "                         sep=\"\\t\", usecols=['chrom', 'start', 'end', 'het-1'])\n",
    "        \n",
    "        df = df.drop_duplicates(subset = [\"chrom\", \"start\",\"end\"])\n",
    "        df['end'] = df['end'] - 1\n",
    "        df = pd.merge(df, old_overlaps_entire, left_on=[\"chrom\", \"start\",\"end\"], \n",
    "                      right_on = [\"chrom\", \"TR_start\",\"TR_end\"])\n",
    "        df = df[['chrom', 'start', 'end', \"Gene_name\", 'het-1']]\n",
    "        all_df[i] = pd.concat([all_df[i], df])\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>EUR_het</th>\n",
       "      <th>SAS_het</th>\n",
       "      <th>EAS_het</th>\n",
       "      <th>AMR_het</th>\n",
       "      <th>AFR_het</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>960770</td>\n",
       "      <td>960798</td>\n",
       "      <td>ENST00000338591.8_cds_0_0_chr1_960694_f</td>\n",
       "      <td>0.00321</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>970841</td>\n",
       "      <td>970857</td>\n",
       "      <td>ENST00000379409.6_cds_4_0_chr1_970686_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>976178</td>\n",
       "      <td>976199</td>\n",
       "      <td>ENST00000433179.4_cds_3_0_chr1_976172_r</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1232633</td>\n",
       "      <td>1232644</td>\n",
       "      <td>ENST00000379198.5_cds_0_0_chr1_1232279_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1232807</td>\n",
       "      <td>1232818</td>\n",
       "      <td>ENST00000379198.5_cds_0_0_chr1_1232279_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6069</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50603316</td>\n",
       "      <td>50603361</td>\n",
       "      <td>ENST00000329492.6_cds_2_0_chr22_50603223_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00341</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6070</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604149</td>\n",
       "      <td>50604160</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6071</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604544</td>\n",
       "      <td>50604573</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6072</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604546</td>\n",
       "      <td>50604557</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6073</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604704</td>\n",
       "      <td>50604741</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6074 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end                                   gene_name  \\\n",
       "0      chr1    960770    960798     ENST00000338591.8_cds_0_0_chr1_960694_f   \n",
       "1      chr1    970841    970857     ENST00000379409.6_cds_4_0_chr1_970686_f   \n",
       "2      chr1    976178    976199     ENST00000433179.4_cds_3_0_chr1_976172_r   \n",
       "3      chr1   1232633   1232644    ENST00000379198.5_cds_0_0_chr1_1232279_f   \n",
       "4      chr1   1232807   1232818    ENST00000379198.5_cds_0_0_chr1_1232279_f   \n",
       "...     ...       ...       ...                                         ...   \n",
       "6069  chr22  50603316  50603361  ENST00000329492.6_cds_2_0_chr22_50603223_f   \n",
       "6070  chr22  50604149  50604160  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "6071  chr22  50604544  50604573  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "6072  chr22  50604546  50604557  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "6073  chr22  50604704  50604741  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "\n",
       "      EUR_het  SAS_het  EAS_het  AMR_het  AFR_het  \n",
       "0     0.00321  0.00000  0.00000      0.0  0.00233  \n",
       "1     0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "2     0.00000  0.00332  0.00000      0.0  0.00000  \n",
       "3     0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "4     0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "...       ...      ...      ...      ...      ...  \n",
       "6069  0.00000  0.00000  0.00341      0.0  0.00447  \n",
       "6070  0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "6071  0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "6072  0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "6073  0.00000  0.00000  0.00000      0.0  0.00000  \n",
       "\n",
       "[6074 rows x 9 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Merging all pops ####\n",
    "\n",
    "df_all_pop = reduce(lambda  left,right: pd.merge(left,right,on=['chrom','start','end', 'Gene_name'],\n",
    "                                            how='inner'), all_df)\n",
    "df_all_pop.columns = ['chrom', 'start', 'end', 'EUR_het', 'gene_name',\n",
    "                      'SAS_het', 'EAS_het', 'AMR_het', 'AFR_het']\n",
    "df_all_pop = df_all_pop[['chrom', 'start', 'end', 'gene_name', 'EUR_het',\n",
    "                      'SAS_het', 'EAS_het', 'AMR_het', 'AFR_het']]\n",
    "df_all_pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>EUR_het</th>\n",
       "      <th>SAS_het</th>\n",
       "      <th>EAS_het</th>\n",
       "      <th>AMR_het</th>\n",
       "      <th>AFR_het</th>\n",
       "      <th>motif</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>960770</td>\n",
       "      <td>960798</td>\n",
       "      <td>ENST00000338591.8_cds_0_0_chr1_960694_f</td>\n",
       "      <td>0.00321</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00233</td>\n",
       "      <td>CCG</td>\n",
       "      <td>[CGCCGCCG]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>970841</td>\n",
       "      <td>970857</td>\n",
       "      <td>ENST00000379409.6_cds_4_0_chr1_970686_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>AC</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>976178</td>\n",
       "      <td>976199</td>\n",
       "      <td>ENST00000433179.4_cds_3_0_chr1_976172_r</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00332</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>AGCCCC</td>\n",
       "      <td>[GCTGGGGCTGGGGCTGGGGCTG, GCTGGGGCTGGGGCTGTGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1232633</td>\n",
       "      <td>1232644</td>\n",
       "      <td>ENST00000379198.5_cds_0_0_chr1_1232279_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>AGC</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1232807</td>\n",
       "      <td>1232818</td>\n",
       "      <td>ENST00000379198.5_cds_0_0_chr1_1232279_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>CCG</td>\n",
       "      <td>[CGCCGC, CGCCGCCGC, CGC]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6069</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50603316</td>\n",
       "      <td>50603361</td>\n",
       "      <td>ENST00000329492.6_cds_2_0_chr22_50603223_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00341</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00447</td>\n",
       "      <td>AGG</td>\n",
       "      <td>[GAGGAGGACGATGAGGACGAGGAAGAGGAGGAGGAGGAGGAGG, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6070</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604149</td>\n",
       "      <td>50604160</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>AGC</td>\n",
       "      <td>[AGC]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6071</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604544</td>\n",
       "      <td>50604573</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>CCGCG</td>\n",
       "      <td>[CGCGCCGCCGCGGCCCGCGCCCGCCGCGCC, CGCGCCGCCGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6072</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604546</td>\n",
       "      <td>50604557</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>CCG</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6073</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50604704</td>\n",
       "      <td>50604741</td>\n",
       "      <td>ENST00000329492.6_cds_4_0_chr22_50603841_f</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>AGG</td>\n",
       "      <td>[GAGGAGGAGGACGAAGGGGAAGACGAGGAAGAGGAGGA, GAGGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6074 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end                                   gene_name  \\\n",
       "0      chr1    960770    960798     ENST00000338591.8_cds_0_0_chr1_960694_f   \n",
       "1      chr1    970841    970857     ENST00000379409.6_cds_4_0_chr1_970686_f   \n",
       "2      chr1    976178    976199     ENST00000433179.4_cds_3_0_chr1_976172_r   \n",
       "3      chr1   1232633   1232644    ENST00000379198.5_cds_0_0_chr1_1232279_f   \n",
       "4      chr1   1232807   1232818    ENST00000379198.5_cds_0_0_chr1_1232279_f   \n",
       "...     ...       ...       ...                                         ...   \n",
       "6069  chr22  50603316  50603361  ENST00000329492.6_cds_2_0_chr22_50603223_f   \n",
       "6070  chr22  50604149  50604160  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "6071  chr22  50604544  50604573  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "6072  chr22  50604546  50604557  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "6073  chr22  50604704  50604741  ENST00000329492.6_cds_4_0_chr22_50603841_f   \n",
       "\n",
       "      EUR_het  SAS_het  EAS_het  AMR_het  AFR_het   motif  \\\n",
       "0     0.00321  0.00000  0.00000      0.0  0.00233     CCG   \n",
       "1     0.00000  0.00000  0.00000      0.0  0.00000      AC   \n",
       "2     0.00000  0.00332  0.00000      0.0  0.00000  AGCCCC   \n",
       "3     0.00000  0.00000  0.00000      0.0  0.00000     AGC   \n",
       "4     0.00000  0.00000  0.00000      0.0  0.00000     CCG   \n",
       "...       ...      ...      ...      ...      ...     ...   \n",
       "6069  0.00000  0.00000  0.00341      0.0  0.00447     AGG   \n",
       "6070  0.00000  0.00000  0.00000      0.0  0.00000     AGC   \n",
       "6071  0.00000  0.00000  0.00000      0.0  0.00000   CCGCG   \n",
       "6072  0.00000  0.00000  0.00000      0.0  0.00000     CCG   \n",
       "6073  0.00000  0.00000  0.00000      0.0  0.00000     AGG   \n",
       "\n",
       "                                                    alt  \n",
       "0                                            [CGCCGCCG]  \n",
       "1                                                    []  \n",
       "2     [GCTGGGGCTGGGGCTGGGGCTG, GCTGGGGCTGGGGCTGTGGCT...  \n",
       "3                                                    []  \n",
       "4                              [CGCCGC, CGCCGCCGC, CGC]  \n",
       "...                                                 ...  \n",
       "6069  [GAGGAGGACGATGAGGACGAGGAAGAGGAGGAGGAGGAGGAGG, ...  \n",
       "6070                                              [AGC]  \n",
       "6071  [CGCGCCGCCGCGGCCCGCGCCCGCCGCGCC, CGCGCCGCCGCCG...  \n",
       "6072                                                 []  \n",
       "6073  [GAGGAGGAGGACGAAGGGGAAGACGAGGAAGAGGAGGA, GAGGA...  \n",
       "\n",
       "[6074 rows x 11 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Polymorphism percentage in coding regions: polymorphic_cnt: 0.5408297662166612\n"
     ]
    }
   ],
   "source": [
    "#### Load Motif and polymorphism ####\n",
    "motifs = []\n",
    "for chrom in range(1,23):\n",
    "    vcf = VCF(f\"/expanse/projects/gymreklab/helia/ensembl/experiments/coding_regions/heterozygosity/coding_calls_{chrom}.vcf.gz\")\n",
    "    for record in vcf:\n",
    "        motifs.append([record.CHROM, record.POS, record.INFO['END'], record.INFO['RU'], record.ALT])\n",
    "\n",
    "motifs = pd.DataFrame(motifs, columns = ['chrom', 'start', 'end', 'motif', 'alt'])\n",
    "motifs = motifs.drop_duplicates(subset = [\"chrom\", \"start\",\"end\"])\n",
    "df_all_pop_motif = pd.merge(df_all_pop, motifs, on = ['chrom', 'start', 'end'])\n",
    "display(df_all_pop_motif)\n",
    "polymorphism_percentage = len(df_all_pop_motif[df_all_pop_motif['alt'].map(len) != 0])/len(df_all_pop_motif)\n",
    "print(f\"Polymorphism percentage in coding regions: polymorphic_cnt: {polymorphism_percentage}\")\n",
    "df_all_pop_motif.drop(labels = \"alt\", axis = 1).to_csv(\"tables/Supplementary_Table_7.csv\", index = False, sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1440104 1123154 0.7799117285973791\n"
     ]
    }
   ],
   "source": [
    "#### Polymorphims in whole genome ####\n",
    "all_records = 0\n",
    "polymorphic = 0\n",
    "for i in range(1,23):\n",
    "    df = pd.read_csv(f\"/expanse/projects/gymreklab/helia/ensembl/experiments/coding_regions/info/stats_chr{i}.txt\",\n",
    "                     sep = \"\\t\", header = None)    \n",
    "    all_records += len(df)\n",
    "    polymorphic += len(df[df[4] != \".\"])\n",
    "    \n",
    "print(all_records, polymorphic, polymorphic/all_records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
