{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 3: Characterizing TR heterozygosity patterns\n",
    "\n",
    "This notebook generates the items listed below. For preprocessing see:\n",
    "* `../het-stats/readme.sh`: computes per-locus, per-population, heterozygosity/num allele stats\n",
    "\n",
    "Figures:\n",
    "* Figure 3a-e: lengths vs. heterozygosity\n",
    "\n",
    "Supplementary Figures:\n",
    "* Supp Figure 14: length vs. het by population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "####### Set up #########\n",
    "\n",
    "%pylab inline\n",
    "\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib.gridspec import GridSpec, GridSpecFromSubplotSpec\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "import seaborn as sns\n",
    "\n",
    "PLOTDIR = \"pdfs\"\n",
    "\n",
    "dfs = []\n",
    "for chrom in range(1, 23):\n",
    "    if not os.path.exists(\"../het-stats/chr%s.period.tab\"%chrom): continue\n",
    "    data = pd.read_csv(\"../het-stats/chr%s.stats.tab\"%chrom, sep=\"\\t\")\n",
    "    dfs.append(data)\n",
    "data = pd.concat(dfs)\n",
    "\n",
    "motifs = pd.read_csv(\"../het-stats/motif_info.tab\", sep=\"\\t\", names=[\"chrom\", \"start\",\"period\",\"motif\"])\n",
    "data = pd.merge(data, motifs, on=[\"chrom\",\"start\"])\n",
    "data[\"reflen\"] = data.apply(lambda x: (x[\"end\"]-x[\"start\"]+1)/x[\"period\"], 1)\n",
    "data[\"reflen-bp\"] = data.apply(lambda x: (x[\"end\"]-x[\"start\"]), 1)\n",
    "\n",
    "data = data[(data[\"motif\"].apply(type)==str)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotate data frame with longest perfect repeat stretch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyfaidx\n",
    "window = 5 # buffer around each STR to check\n",
    "genome = pyfaidx.Fasta(\"/storage/resources/dbase/human/GRCh38/GRCh38_full_analysis_set_plus_decoy_hla.fa\")\n",
    "data[\"refseq\"] = data.apply(lambda x: str(genome[x[\"chrom\"]][x[\"start\"]-window:x[\"end\"]+window]), 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def longest_perfect_repeat(ref, motif):\n",
    "    max_matches = []\n",
    "    strand_ref = \"\"\n",
    "    \n",
    "    for x in ref:\n",
    "        if x == 'T':\n",
    "            strand_ref += \"A\"\n",
    "        if x == \"A\":\n",
    "            strand_ref += \"T\"\n",
    "        if x == \"G\":\n",
    "            strand_ref += \"C\"\n",
    "        if x == \"C\":\n",
    "            strand_ref += \"G\"\n",
    "    strand_ref = strand_ref[::-1]\n",
    "    \n",
    "    for ref_ in [ref, strand_ref]:\n",
    "        for mot in [motif, motif[::-1]]:\n",
    "                i = 0\n",
    "                match = 0\n",
    "                max_match = 0\n",
    "        \n",
    "                while True:\n",
    "                    if i >= len(ref_):\n",
    "                        break\n",
    "                    for j in range(0,len(motif)):\n",
    "                        k = i\n",
    "                        while True:\n",
    "                            while j < len(mot) and k < len(ref_) and ref_[k] == mot[j]:\n",
    "                                k += 1\n",
    "                                j += 1\n",
    "                                match += 1\n",
    "                            max_match = max(max_match, match)\n",
    "                            if j == len(motif):\n",
    "                                j = 0\n",
    "                                i = k\n",
    "                            else:\n",
    "                                if j == len(motif) - 1:\n",
    "                                    i += 1\n",
    "                                match = 0\n",
    "                                break\n",
    "\n",
    "                        j = 0\n",
    "                max_matches.append(max_match)\n",
    "\n",
    "    return max(max_matches)\n",
    "data[\"longest_repeat_stretch\"] = data.apply(lambda x: longest_perfect_repeat(x[\"refseq\"], x[\"motif\"]), 1)\n",
    "data.to_csv(\"data_longest_stretch.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 3, panels a-e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot reflen vs. het by period, overall and for each population\n",
    "LABSIZE = 18\n",
    "TICKSIZE = 14\n",
    "\n",
    "def GetMatchesMotif(motif, motiflist):\n",
    "    for m in motiflist:\n",
    "        if m == \"ALL\": return True\n",
    "        elif type(m) == int or type(m) == float:\n",
    "            if len(motif)==m: return True\n",
    "        elif m == \"vntr\":\n",
    "            if len(motif)> 6: return True\n",
    "        else:\n",
    "            if m == motif: return True\n",
    "    return False\n",
    "    \n",
    "def PlotLenVsHet(data, ax=None, lencol=\"reflen\", lineskip=1, pop=None, motifs=None, colors=None, figname=None):\n",
    "    \"\"\"\n",
    "    Plot length vs. het\n",
    "    If pop=None, use het-ALL. else use het-${pop}\n",
    "    motifs is a list of motifs. or 1, 2, 3 etc. for\n",
    "    all motifs with that length. Use \"vntr\" for 7+\n",
    "    \"\"\"\n",
    "    if lencol == \"reflen\":\n",
    "        MINLEN = 10\n",
    "        MAXLEN = 30\n",
    "    else:\n",
    "        MINLEN = 10\n",
    "        MAXLEN = 80\n",
    "    if pop is None:\n",
    "        col = \"het-ALL\"\n",
    "    else: col = \"het-%s\"%pop\n",
    "    if ax is None:\n",
    "        fig = plt.figure()\n",
    "        fig.set_size_inches((8, 6))\n",
    "        ax = fig.add_subplot(111)\n",
    "    if motifs is None:\n",
    "        motifs = [\"ALL\"]\n",
    "        colors = [\"black\"]\n",
    "    for i in range(len(motifs)):\n",
    "        m = motifs[i]\n",
    "        lenbins = range(MINLEN, MAXLEN+1)\n",
    "        if lencol == \"reflen\":\n",
    "            lenbins = np.arange(MINLEN, MAXLEN, 1)\n",
    "        else:\n",
    "            if (m in [1,2,3] or (type(m)==str and len(m) <3)) and m != \"vntr\" and (m != \"ALL\"):\n",
    "                lenbins = range(MINLEN, MAXLEN+1)\n",
    "            else:\n",
    "                lenbins = np.arange(MINLEN, MAXLEN, 1)  # could go lower resolution here if we want\n",
    "        color = colors[i]\n",
    "        xvals = []\n",
    "        yvals = []\n",
    "        for j in range(len(lenbins)-1):\n",
    "            lb = lenbins[j]\n",
    "            ub = lenbins[j+1]\n",
    "            xx = data[(data[lencol]>=lb) & (data[lencol]<ub) &\n",
    "                                      (data[\"motif\"].apply(lambda x: GetMatchesMotif(x, [m]), 1))]\n",
    "            if xx.shape[0] > 10:\n",
    "                xvals.append(lb)\n",
    "                yvals.append(np.mean(xx[col]))\n",
    "        ax.plot(xvals, yvals, color=color, label=m)\n",
    "    if lineskip is not None:\n",
    "        for i in range(MINLEN, MAXLEN+1, lineskip):\n",
    "            ax.axvline(x=i, linewidth=0.5, color=\"gray\")\n",
    "    if lencol == \"lencol\":\n",
    "        ax.set_xlabel(\"Length\", size=LABSIZE)\n",
    "    else:\n",
    "        ax.set_xlabel(\"Length (bp)\", size=LABSIZE)\n",
    "    ax.set_ylabel(\"Mean heterozygosity\", size=LABSIZE)\n",
    "    ax.legend(loc=\"lower right\")\n",
    "    if pop is not None: ax.set_title(pop)\n",
    "    ax.set_xticklabels([int(item) for item in ax.get_xticks()], size=TICKSIZE)\n",
    "    ax.set_yticklabels([\"%.1f\"%item for item in ax.get_yticks()], size=TICKSIZE)\n",
    "    if figname is not None and ax is None: fig.savefig(figname)\n",
    "        \n",
    "# Plot reflen vs. het by period, overall and for each population\n",
    "\n",
    "def GetMatchesMotif(motif, motiflist):\n",
    "    for m in motiflist:\n",
    "        if m == \"ALL\": return True\n",
    "        elif type(m) == int or type(m) == float:\n",
    "            if len(motif)==m: return True\n",
    "        elif m == \"vntr\":\n",
    "            if len(motif)> 6: return True\n",
    "        else:\n",
    "            if m == motif: return True\n",
    "    return False\n",
    "    \n",
    "def PlotLenVsNum(data, ax=None, lencol=\"reflen\", motifs=None, colors=None, figname=None):\n",
    "    \"\"\"\n",
    "    Plot length vs. num loci\n",
    "    motifs is a list of motifs. or 1, 2, 3 etc. for\n",
    "    all motifs with that length. Use \"vntr\" for 7+\n",
    "    \"\"\"\n",
    "    if lencol == \"reflen\":\n",
    "        MINLEN = 10\n",
    "        MAXLEN = 30\n",
    "    else:\n",
    "        MINLEN = 10\n",
    "        MAXLEN = 80\n",
    "    if ax is None:\n",
    "        fig = plt.figure()\n",
    "        fig.set_size_inches((8, 3))\n",
    "        ax = fig.add_subplot(111)\n",
    "    if motifs is None:\n",
    "        motifs = [\"ALL\"]\n",
    "        colors = [\"black\"]\n",
    "    for i in range(len(motifs)):\n",
    "        m = motifs[i]\n",
    "        lenbins = range(MINLEN, MAXLEN+1)\n",
    "        if lencol == \"reflen\":\n",
    "            lenbins = np.arange(MINLEN, MAXLEN, 1)\n",
    "        else:\n",
    "            if (m in [1,2,3] or (type(m)==str and len(m) <3)) and m != \"vntr\" and (m != \"ALL\"):\n",
    "                lenbins = range(MINLEN, MAXLEN+1)\n",
    "            else:\n",
    "                lenbins = np.arange(MINLEN, MAXLEN, 1)  # could go lower resolution here if we want\n",
    "        color = colors[i]\n",
    "        xvals = []\n",
    "        yvals = []\n",
    "        for j in range(len(lenbins)-1):\n",
    "            lb = lenbins[j]\n",
    "            ub = lenbins[j+1]\n",
    "            xx = data[(data[lencol]>=lb) & (data[lencol]<ub) &\n",
    "                                      (data[\"motif\"].apply(lambda x: GetMatchesMotif(x, [m]), 1))]\n",
    "            \n",
    "            if xx.shape[0] > 10:\n",
    "                xvals.append(lb)\n",
    "                yvals.append(xx.shape[0])\n",
    "        ax.plot(xvals, yvals, color=color, label=m)\n",
    "    if lencol == \"lencol\":\n",
    "        ax.set_xlabel(\"Length\", size=LABSIZE)\n",
    "    else:\n",
    "        ax.set_xlabel(\"Length (bp)\", size=LABSIZE)\n",
    "    ax.set_ylabel(\"Num. TRs\", size=LABSIZE)\n",
    "    plt.yscale(\"log\")\n",
    "    ax.legend(loc=\"lower right\")\n",
    "    ax.set_xticklabels([int(item) for item in ax.get_xticks()], size=TICKSIZE)\n",
    "    ax.set_yticklabels(ax.get_yticklabels(), size=TICKSIZE)\n",
    "    if figname is not None and ax is None: fig.savefig(figname)\n",
    "\n",
    "colors = [\"red\",\"orange\",\"gold\",\"green\",\"blue\",\"purple\",\"pink\",\"brown\",\"cyan\",\"gray\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pltdata = data\n",
    "################## Figure 3 ##############\n",
    "fig = plt.figure()\n",
    "fig.set_size_inches((20, 10))\n",
    "\n",
    "# Overall grid spec\n",
    "gs = GridSpec(2, 3, width_ratios=[1.5,1, 1], height_ratios=[1,1])\n",
    "\n",
    "# Nested grid spec for figures on the left\n",
    "gs00 = GridSpecFromSubplotSpec(2, 1, height_ratios=[1, 2], subplot_spec=gs[:,0])\n",
    "\n",
    "# Plot overall - num\n",
    "ax = fig.add_subplot(gs00[0,0])\n",
    "PlotLenVsNum(pltdata, ax=ax, motifs=[1,2,3,4,5,6,\"vntr\"], \n",
    "             colors=[\"gray\",\"red\",\"gold\",\"blue\",\"green\",\"purple\",\"brown\"],\n",
    "             lencol=\"longest_repeat_stretch\");\n",
    "ax.text(-0.1, 1.05, \"a\", transform=ax.transAxes, \n",
    "                     size=18, fontweight=\"bold\", family=\"sans-serif\",\n",
    "                    va=\"top\", ha=\"right\")\n",
    "\n",
    "# Plot overall - het\n",
    "ax = fig.add_subplot(gs00[1,0])\n",
    "PlotLenVsHet(pltdata, ax=ax, motifs=[1,2,3,4,5,6,\"vntr\"], \n",
    "             colors=[\"gray\",\"red\",\"gold\",\"blue\",\"green\",\"purple\",\"brown\"], lineskip=None,\n",
    "             lencol=\"longest_repeat_stretch\");\n",
    "\n",
    "for period in [2,3,4,5]:\n",
    "    if period == 2:\n",
    "        ax = ax=fig.add_subplot(gs[0,1])\n",
    "        labtext = \"b\"\n",
    "    if period == 3:\n",
    "        ax=fig.add_subplot(gs[0,2])\n",
    "        labtext = \"c\"\n",
    "    if period == 4:\n",
    "        ax=fig.add_subplot(gs[1,1])\n",
    "        labtext = \"d\"\n",
    "    if period == 5:\n",
    "        ax=fig.add_subplot(gs[1,2])\n",
    "        labtext = \"e\"\n",
    "        \n",
    "    motifs = list(pltdata[pltdata[\"period\"]==period].groupby(\"motif\", as_index=False).agg({\"start\": len}).sort_values(\"start\", ascending=False).head(n=10)[\"motif\"])\n",
    "    PlotLenVsHet(pltdata, ax=ax, motifs=motifs, \n",
    "                 colors=colors, \n",
    "                 lineskip=period,\n",
    "             lencol=\"longest_repeat_stretch\");\n",
    "    \n",
    "    ax.text(-0.1, 1.05, labtext, transform=ax.transAxes, \n",
    "                     size=18, fontweight=\"bold\", family=\"sans-serif\",\n",
    "                    va=\"top\", ha=\"right\")\n",
    "\n",
    "fig.tight_layout()\n",
    "fig.savefig(os.path.join(PLOTDIR, \"Figure3a-e.pdf\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supp Figure 14 - length vs. het by population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure()\n",
    "fig.set_size_inches((20, 10))\n",
    "\n",
    "popind = 1\n",
    "for pop in [\"EUR\",\"AMR\",\"SAS\",\"EAS\",\"AFR\",\"H3A\"]:\n",
    "    ax = fig.add_subplot(2, 3, popind)\n",
    "    PlotLenVsHet(pltdata, ax=ax, motifs=[1,2,3,4,5,6,\"vntr\"], pop=pop, lineskip=None, \n",
    "             lencol=\"longest_repeat_stretch\", colors=[\"gray\",\"red\",\"gold\",\"blue\",\"green\",\"purple\",\"brown\"]);          \n",
    "    popind += 1\n",
    "\n",
    "fig.tight_layout()\n",
    "fig.savefig(os.path.join(PLOTDIR, \"SuppFigure14-LengthVsHetByPop.pdf\"));"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
