{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/anaconda3-2020.11-bsn4npoxyw7jzz7fajncek3bvdoaa5wv/lib/python3.8/site-packages/tqdm/std.py:697: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version\n",
      "  from pandas import Panel\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "tqdm.pandas(desc='My bar!')\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/anaconda3-2020.11-bsn4npoxyw7jzz7fajncek3bvdoaa5wv/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    }
   ],
   "source": [
    "genes = pd.read_csv(\"Homo_sapiens.GRCh38.108.gtf\", sep = \"\\t\", comment=\"#\", header=None)\n",
    "genes[0] = \"chr\" + genes[0].astype(str)\n",
    "\n",
    "\n",
    "def find_genes(row):\n",
    "    gene_df = genes[(genes[0] == row['CHROM']) & (genes[3] <= row['POS']) & \n",
    "                    (genes[4] >= row['POS'])]\n",
    "    if len(gene_df) == 0:\n",
    "        return None\n",
    "    gene_df_info = gene_df[8].str.split(\";\")\n",
    "    gene_set = set()\n",
    "    for info in list(gene_df_info):\n",
    "        for field in info:\n",
    "            if \"gene_name\" in field:\n",
    "                name = field.strip().replace(\"gene_name \",\"\").replace('\"','')\n",
    "                gene_set.add(name)\n",
    "    if len(gene_set) == 0:\n",
    "        return None\n",
    "                \n",
    "    return \",\".join(list(gene_set))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_expansions = pd.DataFrame(columns = ['CHROM', 'POS', 'PERIOD', 'MOTIF', 'outlier_threshold', 'AFR_freq',\n",
    "       'NON_AFR_freq'])\n",
    "for i in range(1,23):\n",
    "    if i == 9:\n",
    "        continue\n",
    "    expansions = pd.read_csv(f\"{i}_expansions.csv\")\n",
    "    all_expansions = pd.concat([expansions, all_expansions])\n",
    "    \n",
    "    \n",
    "all_expansions['gene'] = all_expansions.apply(lambda row: find_genes(row), axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "pedigree = pd.read_csv(\"/expanse/projects/gymreklab/helia/TR_1000G/1000G.ped\", delim_whitespace=True)\n",
    "pedigree = pedigree[['SampleID','Superpopulation', 'Population']]\n",
    "\n",
    "samp_to_subpop = pd.Series(pedigree[pedigree['Superpopulation'] == \"AFR\"].Population.values,index=pedigree[pedigree['Superpopulation'] == \"AFR\"].SampleID).to_dict()\n",
    "samp_to_pop = pd.Series(pedigree.Superpopulation.values,index=pedigree.SampleID).to_dict()\n",
    "H3Africa_names = pd.read_csv(\"/expanse/projects/gymreklab/helia/H3Africa/names/H3A_Baylor_sample_country.txt\", header=None, delim_whitespace=True)\n",
    "\n",
    "for index,row in H3Africa_names.iterrows():\n",
    "    samp_to_pop[row[0]] = \"H3Africa\"\n",
    "    samp_to_subpop[row[0]] = row[1]\n",
    "    \n",
    "african_pop = set(list(samp_to_subpop.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>PERIOD</th>\n",
       "      <th>MOTIF</th>\n",
       "      <th>outlier_threshold</th>\n",
       "      <th>AFR_freq</th>\n",
       "      <th>NON_AFR_freq</th>\n",
       "      <th>gene</th>\n",
       "      <th>AFR_freqs</th>\n",
       "      <th>H3Africa_freqs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22</td>\n",
       "      <td>17976536</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.027958</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>MICAL3</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22</td>\n",
       "      <td>20888311</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.077357</td>\n",
       "      <td>0.000866</td>\n",
       "      <td>SNAP29</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr22</td>\n",
       "      <td>22722990</td>\n",
       "      <td>2</td>\n",
       "      <td>AG</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.042406</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>None</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22</td>\n",
       "      <td>24042817</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.032510</td>\n",
       "      <td>0.000650</td>\n",
       "      <td>CABIN1</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22</td>\n",
       "      <td>24496373</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.018623</td>\n",
       "      <td>0.000435</td>\n",
       "      <td>UPB1</td>\n",
       "      <td>0.017</td>\n",
       "      <td>0.023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247419145</td>\n",
       "      <td>2</td>\n",
       "      <td>AT</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.027386</td>\n",
       "      <td>0.000655</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247919144</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.081789</td>\n",
       "      <td>0.001732</td>\n",
       "      <td>None</td>\n",
       "      <td>0.077</td>\n",
       "      <td>0.093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248249150</td>\n",
       "      <td>2</td>\n",
       "      <td>AT</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.055197</td>\n",
       "      <td>0.001083</td>\n",
       "      <td>None</td>\n",
       "      <td>0.058</td>\n",
       "      <td>0.047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248684754</td>\n",
       "      <td>2</td>\n",
       "      <td>AT</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.072522</td>\n",
       "      <td>0.002599</td>\n",
       "      <td>None</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248874904</td>\n",
       "      <td>4</td>\n",
       "      <td>AGAT</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.021757</td>\n",
       "      <td>0.000650</td>\n",
       "      <td>None</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2055 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM        POS PERIOD MOTIF  outlier_threshold  AFR_freq  NON_AFR_freq  \\\n",
       "0    chr22   17976536      2    AC                6.0  0.027958      0.000000   \n",
       "1    chr22   20888311      2    AC                8.0  0.077357      0.000866   \n",
       "2    chr22   22722990      2    AG                9.0  0.042406      0.000000   \n",
       "3    chr22   24042817      2    AC               16.0  0.032510      0.000650   \n",
       "4    chr22   24496373      2    AC                7.0  0.018623      0.000435   \n",
       "..     ...        ...    ...   ...                ...       ...           ...   \n",
       "173   chr1  247419145      2    AT                4.5  0.027386      0.000655   \n",
       "174   chr1  247919144      2    AC                5.0  0.081789      0.001732   \n",
       "175   chr1  248249150      2    AT                8.0  0.055197      0.001083   \n",
       "176   chr1  248684754      2    AT                6.0  0.072522      0.002599   \n",
       "177   chr1  248874904      4  AGAT                6.0  0.021757      0.000650   \n",
       "\n",
       "       gene  AFR_freqs  H3Africa_freqs  \n",
       "0    MICAL3      0.028           0.029  \n",
       "1    SNAP29      0.071           0.093  \n",
       "2      None      0.049           0.039  \n",
       "3    CABIN1      0.036           0.024  \n",
       "4      UPB1      0.017           0.023  \n",
       "..      ...        ...             ...  \n",
       "173   NLRP3      0.030           0.021  \n",
       "174    None      0.077           0.093  \n",
       "175    None      0.058           0.047  \n",
       "176    None      0.074           0.069  \n",
       "177    None      0.022           0.022  \n",
       "\n",
       "[2055 rows x 10 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_freqs(row):\n",
    "    chrom = row['CHROM'].replace(\"chr\", \"\")\n",
    "    header = []\n",
    "    with open(f\"/expanse/projects/gymreklab/helia/ensembl/experiments/charact/diff_dist/diff/{chrom}_diff.txt\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"chr\"):\n",
    "                    f.close()\n",
    "                    break\n",
    "            header.append(line.strip())\n",
    "\n",
    "    addr=f\"/expanse/projects/gymreklab/helia/ensembl/experiments/charact/diff_dist/diff/{chrom}_diff.txt\"\n",
    "\n",
    "    pos = row['POS']\n",
    "    x=!grep -m 1 -w $pos $addr\n",
    "    gbs = x[0].split(\"\\t\")[4:]\n",
    "    dict_ = dict(zip(header, gbs))\n",
    "    \n",
    "    for x in dict_:\n",
    "        dict_[x] = dict_[x].split(\"/\")\n",
    "        modified_gbs = []\n",
    "        for s in dict_[x]:\n",
    "            if s != \".\" and s != \"\":\n",
    "                modified_gbs.append(int(s)/ row['PERIOD'])\n",
    "                \n",
    "        dict_[x] = modified_gbs\n",
    "        \n",
    "        \n",
    "        \n",
    "    pop_freqs = {}\n",
    "    for pop in ['AFR', 'H3Africa']:\n",
    "        gbs = []\n",
    "        for samp in dict_:\n",
    "            if samp in samp_to_pop:\n",
    "                if samp_to_pop[samp] == pop:\n",
    "                    gbs.extend(dict_[samp])\n",
    "                \n",
    "        if len(gbs) == 0:\n",
    "            pop_freqs[pop] = -1\n",
    "        else:\n",
    "            above_threshold = round(len([x for x in gbs if x not in [\".\",\"\"] \\\n",
    "                       and x > row['outlier_threshold']]) / len([x for x in gbs if x not in [\".\",\"\"]]),3)\n",
    "            pop_freqs[pop] = above_threshold\n",
    "                \n",
    "                \n",
    "            \n",
    "    \n",
    "    \n",
    "    \n",
    "    return pop_freqs\n",
    "\n",
    "all_expansions[['AFR_freqs', \n",
    "                        'H3Africa_freqs']] = all_expansions.apply(lambda row: find_freqs(row), \n",
    "                                                                         axis = 1, result_type=\"expand\")\n",
    "all_expansions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-22-48b6c8015c9b>:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  all_expansions_both['outlier_threshold'] = all_expansions_both['outlier_threshold'].round(1)\n",
      "<ipython-input-22-48b6c8015c9b>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  all_expansions_both['NON_AFR_freq'] = all_expansions_both['NON_AFR_freq'].round(3)\n",
      "<ipython-input-22-48b6c8015c9b>:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  all_expansions_both['AFR_freq'] = all_expansions_both['AFR_freq'].round(3)\n"
     ]
    }
   ],
   "source": [
    "all_expansions_both = all_expansions[(all_expansions['AFR_freq'] < 0.01) | \n",
    "                                     (all_expansions['AFR_freqs'] > 0.01) & \n",
    "                                     (all_expansions['H3Africa_freqs'] > 0.01)]\n",
    "all_expansions_both['outlier_threshold'] = all_expansions_both['outlier_threshold'].round(1)\n",
    "all_expansions_both['NON_AFR_freq'] = all_expansions_both['NON_AFR_freq'].round(3)\n",
    "all_expansions_both['AFR_freq'] = all_expansions_both['AFR_freq'].round(3)\n",
    "all_expansions_both.columns = ['CHROM','POS','PERIOD','MOTIF','OUTLIER_THRESH',\n",
    "                               'ALL_AFR_FREQ','NON_AFR_FREQ','GENE','1KG_AFR_FREQ','H3Africa_FREQ']\n",
    "all_expansions_both = all_expansions_both[all_expansions_both['OUTLIER_THRESH'] > 10]\n",
    "#all_expansions_both.sort_values('OUTLIER_THRESH', ascending=False).to_csv('Large_Expansions.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Different motif at locus 92185591\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>PERIOD</th>\n",
       "      <th>MOTIF</th>\n",
       "      <th>OUTLIER_THRESH</th>\n",
       "      <th>ALL_AFR_FREQ</th>\n",
       "      <th>NON_AFR_FREQ</th>\n",
       "      <th>GENE</th>\n",
       "      <th>1KG_AFR_FREQ</th>\n",
       "      <th>H3Africa_FREQ</th>\n",
       "      <th>EH_supported</th>\n",
       "      <th>STRetch_supported</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22</td>\n",
       "      <td>24042817</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.001</td>\n",
       "      <td>CABIN1</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.024</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr22</td>\n",
       "      <td>24757660</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.001</td>\n",
       "      <td>PIWIL3</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.011</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr22</td>\n",
       "      <td>26667333</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.001</td>\n",
       "      <td>MIAT</td>\n",
       "      <td>0.078</td>\n",
       "      <td>0.089</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr22</td>\n",
       "      <td>39686600</td>\n",
       "      <td>2</td>\n",
       "      <td>AT</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.037</td>\n",
       "      <td>CACNA1I</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.001</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr22</td>\n",
       "      <td>41958483</td>\n",
       "      <td>2</td>\n",
       "      <td>AG</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.002</td>\n",
       "      <td>SMIM45</td>\n",
       "      <td>0.076</td>\n",
       "      <td>0.055</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>chr1</td>\n",
       "      <td>227297941</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.001</td>\n",
       "      <td>CDC42BPA</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.029</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>chr1</td>\n",
       "      <td>227297968</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.001</td>\n",
       "      <td>CDC42BPA</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.032</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>235882040</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.000</td>\n",
       "      <td>LYST</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.029</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>chr1</td>\n",
       "      <td>244553603</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.044</td>\n",
       "      <td>0.001</td>\n",
       "      <td>CATSPERE</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.056</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245088315</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.001</td>\n",
       "      <td>EFCAB2</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.039</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM        POS PERIOD MOTIF  OUTLIER_THRESH  ALL_AFR_FREQ  \\\n",
       "3    chr22   24042817      2    AC            16.0         0.033   \n",
       "5    chr22   24757660      2    AC            16.0         0.013   \n",
       "6    chr22   26667333      2    AC            11.0         0.081   \n",
       "16   chr22   39686600      2    AT            11.0         0.003   \n",
       "18   chr22   41958483      2    AG            12.0         0.070   \n",
       "..     ...        ...    ...   ...             ...           ...   \n",
       "148   chr1  227297941      2    AC            12.0         0.028   \n",
       "149   chr1  227297968      2    AC            12.0         0.031   \n",
       "159   chr1  235882040      2    AC            12.0         0.027   \n",
       "167   chr1  244553603      2    AC            13.0         0.044   \n",
       "168   chr1  245088315      2    AC            14.0         0.029   \n",
       "\n",
       "     NON_AFR_FREQ      GENE  1KG_AFR_FREQ  H3Africa_FREQ EH_supported  \\\n",
       "3           0.001    CABIN1         0.036          0.024            N   \n",
       "5           0.001    PIWIL3         0.013          0.011            N   \n",
       "6           0.001      MIAT         0.078          0.089            N   \n",
       "16          0.037   CACNA1I         0.003          0.001            N   \n",
       "18          0.002    SMIM45         0.076          0.055            N   \n",
       "..            ...       ...           ...            ...          ...   \n",
       "148         0.001  CDC42BPA         0.027          0.029            N   \n",
       "149         0.001  CDC42BPA         0.030          0.032            N   \n",
       "159         0.000      LYST         0.027          0.029            N   \n",
       "167         0.001  CATSPERE         0.040          0.056            N   \n",
       "168         0.001    EFCAB2         0.026          0.039            N   \n",
       "\n",
       "    STRetch_supported  \n",
       "3                   Y  \n",
       "5                   N  \n",
       "6                   N  \n",
       "16                  N  \n",
       "18                  N  \n",
       "..                ...  \n",
       "148                 N  \n",
       "149                 N  \n",
       "159                 N  \n",
       "167                 N  \n",
       "168                 N  \n",
       "\n",
       "[246 rows x 12 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### Compare with Ibra's results on ExpansionHunter Denovo and STRetch ####\n",
    "\n",
    "\n",
    "EH = pd.read_csv(\"/expanse/projects/gymreklab/helia/ensembl/experiments/charact/diff_dist/STRetch-EHdn/EHdn/loci_detected_by_EHdn.csv\")\n",
    "stretch = pd.read_csv(\"/expanse/projects/gymreklab/helia/ensembl/experiments/charact/diff_dist/STRetch-EHdn/STRetch/TRs_detected_by_STRetch.csv\")\n",
    "stretch.columns = ['chr', 'start', 'end', 'motif', 'gene', 'dist']\n",
    "\n",
    "def support(row, method):\n",
    "    if method == \"EH\":\n",
    "        compare_df = EH.copy()\n",
    "    if method == \"stretch\":\n",
    "        compare_df = stretch.copy()\n",
    "\n",
    "    df = compare_df[(compare_df['chr'] == row['CHROM']) & \n",
    "                    (compare_df['start'] >= row['POS'] - 1000 ) & (compare_df['start'] < row['POS'] + 1000)]\n",
    "    if len(df) == 0:\n",
    "        return \"N\"\n",
    "    if len(df) > 1:\n",
    "        cnt = 0\n",
    "        for index, row1 in df.iterrows():\n",
    "            if row1['motif'] == row['MOTIF']:\n",
    "                cnt += 1\n",
    "        if cnt == 1:\n",
    "            return \"Y\"\n",
    "        else:\n",
    "            print(f\"Multiple records at locus {row['POS']}\")\n",
    "            display(df)\n",
    "            return \"U\"\n",
    "    if len(df) == 1 and str(df['motif'].iloc[0]) == str(row['MOTIF']):\n",
    "        return \"Y\"\n",
    "    if len(df) == 1 and str(df['motif'].iloc[0]) != str(row['MOTIF']):\n",
    "        print(f\"Different motif at locus {row['POS']}\")\n",
    "        return \"U\"\n",
    "    else:\n",
    "        display(df)\n",
    "\n",
    "\n",
    "all_expansions_both['EH_supported'] = all_expansions_both.apply(lambda x: support(x, \"EH\"), axis = 1)\n",
    "all_expansions_both['STRetch_supported'] = all_expansions_both.apply(lambda x: support(x, \"stretch\"), axis = 1)\n",
    "all_expansions_both"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chr', 'start', 'end', 'repeatunit', 'gene', 'dist'], dtype='object')"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stretch.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
